### ISSUE TITLE
Cygfuse fs: persistent disconnect problems

### ISSUE THREAD

**Issue by hasse69** [billziss-gh/winfsp#90]
Even after all we did in https://github.com/hasse69/rar2fs/issues/75 regarding the disconnect issues from explorer the problem is back again :( I am not sure what has happened but now disconnect from the explorer does not work anymore. I mount through the explorer but when I disconnect the drive stays in the list. However, if I try to disconnect again, it says the drive does not exist!? So it disconnected at some level but not all the way it seems. The drive can still be accessed and the process serving the file system is still running. The only way to disconnect it is to terminate the process from the task manager which also leaves me with a stale red crossed drive....
I have completely uninstalled Dokany and also reinstalled WinFSP B3, no change.
I have doubled checked that mounting manually and sending the process a SIGTERM in fact terminates the process. I have so far only tested on a 32-bit system. Maybe this can also be related to my other problems with 32-bit Windows?

**Comment by billziss-gh** [comment]
Disconnect works as follows:

- Explorer instructs the Network Provider DLL to "cancel the connection" by calling [`NPCancelConnection`](https://github.com/billziss-gh/winfsp/blob/v1.0/src/dll/np.c#L765).

- `NPCancelConnection` connects to the WinFsp.Launcher service and sends it a `stop` command for the file system in question. [[link](https://github.com/billziss-gh/winfsp/blob/v1.0/src/dll/np.c#L801)]

- The WinFsp.Launcher service sends a [`CTRL_BREAK_EVENT`](https://github.com/billziss-gh/winfsp/blob/v1.0/src/launcher/launcher.c#L80) to the associated file system process.

- The file system process handles the CTRL_BREAK_EVENT in [`FspServiceConsoleCtrlHandler`](https://github.com/billziss-gh/winfsp/blob/v1.0/src/dll/service.c#L512) by terminating its "service" loop (and `fuse_loop` in case of FUSE).

To debug this procedure we can use the `launchctl` utility.

1. The `list` command should tell you what file system instances currently exist. If you launch a file system with `\\rar2fs\share\path` you should see a file system with "class" name `rar2fs` and "instance" name `share\path`.

    ```
    $ ./launchctl-x64.exe list
    OK
    sshfs billziss@macbook-pro
    ```

2. The `stop` command should stop the file system and report `OK`:

    ```
    $ launchctl-x64.exe stop sshfs billziss@macbook-pro
    OK
    $ launchctl-x64.exe list
    OK
    ```

Do these commands work for you?

An additional issue is that Windows Explorer appears to have a flaw. When a drive gets disconnected the right pane of Explorer gets immediately notified, but the left pane may still show the drive connected for a while. Try to refresh the Explorer (key `F5`) and see if the drive goes away.

**Comment by hasse69** [comment]
> An additional issue is that Windows Explorer appears to have a flaw. When a drive gets disconnected the right pane of Explorer gets immediately notified, but the left pane may still show the drive connected for a while. Try to refresh the Explorer (key F5) and see if the drive goes away.

As stated the drive is still accessible, I can open it and browse its contents so this is not a "refresh" issue only.

**Comment by billziss-gh** [comment]
> As stated the drive is still accessible, I can open it and browse its contents so this is not a "refresh" issue only.

That is somewhat curious because the launcher terminates the file system after 5 seconds forcefully. I suspect that there is something off with the Network Provider DLL or the Provider Order again. Please double check the order as well.

**Comment by hasse69** [comment]
YEA! That was it. Funny that it worked before. Must have been lucky :)
You can close this issue now.

**Comment by billziss-gh** [comment]
No problem. Closing this. Check if this also resolves #89.
